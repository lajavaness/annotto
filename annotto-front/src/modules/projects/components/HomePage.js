import { Button } from 'antd'
import { isEmpty } from 'lodash'
import { useDispatch, useSelector } from 'react-redux'
import { useTranslation } from 'react-i18next'
import React, { useEffect, useMemo, useRef } from 'react'
import dayjs from 'dayjs'
import relativeTime from 'dayjs/plugin/relativeTime'

import { CREATE_PROJECT } from 'shared/enums/privilegesTypes'
import { PROJECTS_SIZE } from 'shared/enums/projectsTypes'

import { fetchProjects } from 'modules/projects/actions/projectsActions'

import usePrivileges from 'shared/hooks/usePrivileges'

import {
	selectIndexProjects,
	selectIsReady,
	selectProjects,
	selectTotalProjects,
} from 'modules/projects/selectors/projectsSelectors'
import { selectProjectClients } from 'modules/root/selectors/rootSelectors'

import EllipsisTableTooltip from 'shared/components/table/EllipsisTableTooltip'
import Loader from 'shared/components/loader/Loader'
import Table from 'shared/components/table/Table'

import * as Styled from 'modules/projects/components/__styles__/HomePage.styles'

const HomePage = ({ setHeaderActions }) => {
	const tableRef = useRef(null)

	const isReady = useSelector(selectIsReady)
	const projects = useSelector(selectProjects)
	const index = useSelector(selectIndexProjects)
	const total = useSelector(selectTotalProjects)
	const clients = useSelector(selectProjectClients)

	const dispatch = useDispatch()

	const { t } = useTranslation(['root', 'projects'])

	const hasPrivilege = usePrivileges()

	dayjs.extend(relativeTime)

	const columns = [
		{
			title: t('projects:table.name'),
			fixed: 'left',
			textWrap: 'word-break',
			ellipsis: true,
			onCell: () => {
				return {
					style: {
						whiteSpace: 'nowrap',
						maxWidth: 175,
					},
				}
			},
			// eslint-disable-next-line react/display-name
			render: ({ _id, name }) => (
				<EllipsisTableTooltip title={name}>
					<Styled.Link to={`/project/${_id}`}>{name}</Styled.Link>,
				</EllipsisTableTooltip>
			),
		},
		{
			title: <Styled.TextRight>{t('projects:table.client')}</Styled.TextRight>,
			dataIndex: 'client',
			width: 100,
			textWrap: 'word-break',
			ellipsis: true,
			// eslint-disable-next-line react/display-name
			render: ({ name }) => <Styled.TextRight>{name}</Styled.TextRight>,
		},
		{
			title: <Styled.TextRight>{t('projects:table.users')}</Styled.TextRight>,
			dataIndex: 'users',
			width: 70,
			textWrap: 'word-break',
			ellipsis: true,
			// eslint-disable-next-line react/display-name
			render: (data) => <Styled.TextRight>{data}</Styled.TextRight>,
		},
		{
			title: <Styled.TextRight>{t('projects:table.items')}</Styled.TextRight>,
			dataIndex: 'items',
			width: 100,
			textWrap: 'word-break',
			ellipsis: true,
			// eslint-disable-next-line react/display-name
			render: (data) => <Styled.TextRight>{data}</Styled.TextRight>,
		},
		{
			title: <Styled.TextRight>{t('projects:table.progress')}</Styled.TextRight>,
			dataIndex: 'progress',
			width: 100,
			textWrap: 'word-break',
			ellipsis: true,
			// eslint-disable-next-line react/display-name
			render: (data) => <Styled.TextRight>{data && t('root:units.percentLabel', { count: data })}</Styled.TextRight>,
		},
		{
			title: <Styled.TextRight>{t('projects:table.velocity')}</Styled.TextRight>,
			dataIndex: 'velocity',
			width: 100,
			onCell: () => {
				return {
					style: {
						whiteSpace: 'nowrap',
						maxWidth: 100,
					},
				}
			},
			textWrap: 'word-break',
			ellipsis: true,
			// eslint-disable-next-line react/display-name
			render: (data) => <Styled.TextRight>{data && t('root:units.secondLabel', { count: data })}</Styled.TextRight>,
		},
		{
			title: <Styled.TextRight>{t('projects:table.remaining_time')}</Styled.TextRight>,
			dataIndex: 'remaining_time',
			width: 150,
			// eslint-disable-next-line react/display-name
			render: (date) => (
				<Styled.TextRight>
					{date && t('root:units.hourLabel', { count: dayjs(date).diff(dayjs(), 'hour') })}
				</Styled.TextRight>
			),
		},
		{
			title: <Styled.TextRight>{t('projects:table.remaining_work')}</Styled.TextRight>,
			dataIndex: 'remaining_work',
			textWrap: 'word-break',
			ellipsis: true,
			width: 150,
			// eslint-disable-next-line react/display-name
			render: (data) => <Styled.TextRight>{data && t('root:units.hourLabel', { count: data })}</Styled.TextRight>,
		},
		{
			title: <Styled.TextRight>{t('projects:table.notes')}</Styled.TextRight>,
			dataIndex: 'notes',
			textWrap: 'word-break',
			ellipsis: true,
			width: 70,
			// eslint-disable-next-line react/display-name
			render: (data) => <Styled.TextRight>{data}</Styled.TextRight>,
		},
		{ title: t('projects:table.update'), dataIndex: 'update', render: (date) => date && dayjs(date).fromNow() },
		{
			width: 100,
			fixed: 'right',
			// eslint-disable-next-line react/display-name
			render: (id) => <Styled.Link to={`/project/${id}/annotation`}>{t('projects:table.annotate')}</Styled.Link>,
			dataIndex: '_id',
		},
	]

	const data = useMemo(
		() =>
			projects?.map(
				(
					{
						_id,
						name,
						client: clientId,
						users,
						itemCount,
						progress,
						velocity,
						remainingWork,
						deadline,
						commentCount,
						updatedAt,
					},
					index
				) => ({
					key: index,
					name,
					client: clients.data.find(({ _id }) => _id === clientId),
					users: !isEmpty(users) ? users.length : 0,
					items: itemCount,
					progress,
					velocity,
					remaining_time: deadline,
					remaining_work: remainingWork,
					notes: commentCount,
					update: updatedAt,
					_id,
				})
			) || [],
		[projects, clients]
	)

	const _onLoadingMoreClick = () => dispatch(fetchProjects(index + 1, PROJECTS_SIZE))

	useEffect(() => {
		if (setHeaderActions) {
			if (hasPrivilege(CREATE_PROJECT)) {
				setHeaderActions(
					<Styled.CreateButtonLink className="ant-btn ant-btn-primary" to={'/create_project/config'}>
						<span>{t('projects:createButton')}</span>
					</Styled.CreateButtonLink>
				)
			} else {
				setHeaderActions(null)
			}
		}
	}, [setHeaderActions, hasPrivilege, t])

	return isReady ? (
		<Styled.Root>
			<Styled.TableContainer ref={tableRef}>
				<Table
					columns={columns}
					dataSource={data}
					footer={
						total !== projects.length &&
						(() => <Button onClick={_onLoadingMoreClick}>{t(`projects:button.loadMore`)}</Button>)
					}
				/>
			</Styled.TableContainer>
		</Styled.Root>
	) : (
		<Loader />
	)
}

export default HomePage
