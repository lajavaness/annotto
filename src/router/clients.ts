import { Router } from 'express'
import auth from '../auth/service'
import clients from '../middleware/clients'
import { schemaValidationMiddleware } from '../middleware/schema-validation-middleware'
import { createClientSchema } from './validation/client'

const init = (router: Router) => {
  /**
   * @typedef ClientPayload
   * @property {string} name.required - Client name.
   * @property {string} description - Client Description.
   * @property {boolean} isActive - Client status.
   */

  /**
   * @typedef Client
   * @property {string} name - Client name.
   * @property {string} description - Client Description.
   * @property {string} id - Client id.
   * @property {boolean} isActive - Client status.
   */

  /**
   * List clients ( realmRole : dataScientist ) - Check List Model for pagination details.
   * @route GET /clients
   * @group clients
   * @security Bearer
   * @param {string} index.query Page index to fetch ( Pages are arrays of data with a length set by limit ).
   * @param {string} limit.query Page length ( default : 10 ). *.
   * @returns {List.model} 200 - Success.
   * @returns {AnnottoError.model}  401 - Unauthorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 - Server Error.
   */
  router.get('/clients', auth.isAuthenticated({}), clients.index)

  /**
   * Get a specific client  ( realmRole : dataScientist ).
   * @route GET /clients/:clientId
   * @group clients
   * @security Bearer
   * @param {string} clientId.path.required Client id.
   * @returns {Client.model} 200 - Success.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  404 - Client not found.
   * @returns {AnnottoError.model}  500 - Fail to get client.
   */

  router.get('/clients/:clientId', auth.isAuthenticated({ realmRole: 'dataScientist' }), clients.getById)

  /**
   * Create client ( realmRole : dataScientist ).
   * @route POST /clients
   * @group clients
   * @security Bearer
   * @param {ClientPayload.model} ClientPayload.body.required .
   * @returns {Client.model} 200 - Success.
   * @returns {AnnottoError.model}  400 - Invalid Payload.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 - Fail to create client.
   */

  router.post(
    '/clients',
    auth.isAuthenticated({ realmRole: 'dataScientist' }),
    schemaValidationMiddleware({ schemaBody: createClientSchema }),
    clients.create
  )

  /**
   * Update a specific client  ( realmRole : dataScientist ).
   * @route PUT /clients/:clientId
   * @param {string} clientId.path.required Client id.
   * @param {ClientPayload.model} ClientPayload.body.required .
   * @group clients
   * @security Bearer
   * @returns {Client.model} 200 - Success.
   * @returns {AnnottoError.model}  400 - Invalid Payload.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  404 - Client not found.
   * @returns {AnnottoError.model}  500 - Fail to update client.
   */

  router.put(
    '/clients/:clientId',
    auth.isAuthenticated({ realmRole: 'dataScientist' }),
    schemaValidationMiddleware({ schemaBody: createClientSchema }),
    clients.update
  )

  /**
   * Delete a specific client  ( realmRole : admin ).
   * @route DELETE /clients/:clientId
   * @param {string} clientId.path.required Client id.
   * @group clients
   * @security Bearer
   * @returns {} 200 - Success.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  404 - Client not found.
   * @returns {AnnottoError.model}  500 - Fail to delete client.
   */

  router.delete('/clients/:clientId', auth.isAuthenticated({ realmRole: 'admin' }), clients.destroy)
}

export default init
