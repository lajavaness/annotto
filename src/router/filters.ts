import { Router } from 'express'
import auth from '../auth/service'
import filters from '../middleware/filters'
import { schemaValidationMiddleware } from '../middleware/schema-validation-middleware'
import { filterBodySchema } from './validation/filter'

/**
 * @typedef Filter
 * @property { string } _id - Id of the filter.
 * @property { string } project - Id of the project associated.
 * @property { string } type - Type of the data.
 * @property { object } payload - Req.body of the filter creation.
 * @property { object } user - User that created filter.
 * @property { string } criteria - Actual query to be applied.
 */

/**
 * @typedef field
 * @property { Array.<string> } operators
 */
/**
 * @typedef filterOperators
 * @property { string } name - Operator name.
 * @property { object } param - Value format associated with operator.
 */
/**
 * @typedef filterFields
 * @property { field.model } fieldName
 */
/**
 * @typedef filterSpecs
 * @property { Array.<filterOperators> } operators
 * @property { Array.<filterFields> } fields
 */

/**
 * @typedef FilterPayload
 * @property { string } operator.required - Operator to apply.
 * @property { string } field.required - Field or Alias for DB path to apply operator on.
 * @property { string } value.required - Value to use for operator.
 */

/**
 * Initialize filter routes.
 * @param {*} router The router.
 */
const init = (router: Router) => {
  /**
   * Create a filter ( minimum project role : user ) - GET /filter/operators for FilterPayload possibilities
   * See confluence Doc for more details : https://lajavaness.atlassian.net/wiki/spaces/AN/pages/2088828992/Filtres.
   * @route POST /projects/:projectId/filter
   * @group filters
   * @security Bearer
   * @param {Array.<FilterPayload>} FilterPayloads.body.required
   * @returns {Filter.model} 200 - Success.
   * @returns {AnnottoError.model} 400 - Invalid Payload or ERROR_FILTER_EXIST_FOR_PROJECT_AND_USER.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 - Server Error.
   */
  router.post(
    '/projects/:projectId/filter',
    auth.isAuthenticated({ projectRole: 'user' }),
    schemaValidationMiddleware({ schemaBody: filterBodySchema }),
    filters.create
  )

  /**
   * Get a user's filter for a project ( minimum project role : user ).
   * @route GET /projects/:projectId/filter
   * @group filters
   * @security Bearer
   * @returns {Filter.model} 200 - Success.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 - Server Error.
   */
  router.get('/projects/:projectId/filter', auth.isAuthenticated({ projectRole: 'user' }), filters.getByProjectAndUser)

  /**
   * Update a user's filter for a project ( minimum project role : user ).
   * @route PUT /projects/:projectId/filter/:filterId
   * @group filters
   * @security Bearer
   * @returns {Filter.model}  200 - Success.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 - Server Error.
   */
  router.put(
    '/projects/:projectId/filter',
    auth.isAuthenticated({ projectRole: 'user' }),
    schemaValidationMiddleware({ schemaBody: filterBodySchema }),
    filters.update
  )

  /**
   * Delete a user's filter for a project ( minimum project role : user ).
   * @route DELETE /projects/:projectId/filter
   * @group filters
   * @security Bearer
   * @returns { object }  200 - Success.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 - Server Error.
   */

  router.delete('/projects/:projectId/filter', auth.isAuthenticated({ projectRole: 'user' }), filters.delete)

  /**
   * Get Specs for filter creation ( minimum project role : user ).
   *
   * Returns : fields array : ( possible operators to apply on each field ) and
   * operators array ( value property format for each operator ).
   *
   * For ex the format for "annotated" + "equal" is : { "operator": "equal", "field": "annotated", value: true }.
   * @route GET /projects/:projectId/filter/operators
   * @group filters
   * @security Bearer
   * @returns { filterSpecs.model }  200 - Success.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 - Server Error.
   */

  router.get(
    '/projects/:projectId/filter/operators',
    auth.isAuthenticated({ projectRole: 'user' }),
    filters.getOperators
  )
}

export default init
