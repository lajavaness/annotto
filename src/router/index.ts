import express from 'express'
import control from '../middleware/control'
import addUsersRoutes from './users'
import addProjectsRoutes from './projects'
import addItemsRoutes from './items'
import addClientsRoutes from './clients'
import addFilterRoutes from './filters'
import addProfilesRoutes from './profiles'
import { keycloak } from '../auth'

const apiRouter = () => {
  const r = express.Router()

  /**
   * Get health status.
   * @route GET /health
   * @group health-api
   * @security Bearer
   * @returns {Health.model} 200 - Success.
   * @returns {AnnottoError.model} 400 - Invalid Credentials Error.
   * @returns {AnnottoError.model} 500 - Place get error.
   */
  r.get('/health', control.health)

  // After this line, every routes are keycloak dependant.
  r.use(keycloak.middleware())

  addUsersRoutes(r)
  addProjectsRoutes(r)
  addItemsRoutes(r)
  addClientsRoutes(r)
  addFilterRoutes(r)
  addProfilesRoutes(r)

  /**
   * @typedef List
   * @property {number} count Page length ( to set page max length to other value than default 10, use limit query param ).
   * @property {number} index Index of the page ( to access specific index use index query param ).
   * @property {number} pageCount Number of Page ( data length / page length ).
   * @property {number} total Total data length.
   * @property {Array.<object>} data
   */

  /**
   * @typedef UserExtern
   * @property {string} _id _id.
   * @property {string} firstName FirstName of the item.
   * @property {string} lastName LastName of the user.
   * @property {string} email Email.
   */

  /**
   * @typedef AnnottoError
   * @property {number} code Error code - eg: 401.
   * @property {string} message - Error message.
   */

  /**
   * @typedef Health
   * @property {string} status Status of the server.
   * @property {string} version Current version of the API.
   * @property {string} environment Current environment of the API.
   */

  /**
   * @typedef User
   * @property {string} firstName FirstName of the user.
   * @property {string} lastName LastName of the user.
   * @property {number} userId Id of the user.
   * @property {boolean} isActive User can login or not.
   * @property {string} email Email address of the user.
   */

  /**
   * @typedef UserUpdate
   * @property {string} firstName FirstName of the user.
   * @property {string} lastName LastName of the user.
   */

  /**
   * @typedef Auth
   * @property{string} email
   * @property {string} password Password.
   */

  /**
   * @typedef LoginCred
   * @property {number} userId Id of the user.
   * @property {string} token Token to provide.
   */

  return r
}

export default apiRouter
