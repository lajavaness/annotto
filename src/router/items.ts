import { Router } from 'express'
import auth from '../auth/service'
import items from '../middleware/items'
import { schemaValidationMiddleware } from '../middleware/schema-validation-middleware'
import { updateItemSchema, annotationPayloadSchema } from './validation/item'
/**
 * @typedef Item
 * @property { string } _id - Id of the items.
 * @property { string } type - Type of the project.
 * @property { string } uuuid - Data uuid.
 * @property { string } type - Type of the data.
 * @property { boolean } annotated - Data already annotated.
 * @property { Array.<string> } tags - List of tags assigned to the data.
 * @property { string } status - Status of the data.
 * @property { object } predictions - Predictions of tasks.
 * @property { string } description - Description of the data.
 */

/**
 * @typedef ItemPayload
 * @property {Array.<string>} tags - List of tags to assign to the item.
 */

/**
 * @typedef Zone
 * @property {number} x - X.
 * @property {number} y - Y.
 */

/**
 * @typedef Ner
 * @property {number} start - Start index.
 * @property {number} end - End index.
 */

/**
 * @typedef Relation
 * @property {object} src - Annotation object : source task.
 * @property {object} dest - Annotation object : destination task.
 * @property {string} value - Value of the relation between the two.
 */

/**
 * @typedef Annotation
 * @property {string} value - Value of the annotation.
 * @property {Array.<Zone>} zone - Zone to annotate.
 * @property {Ner.model} ner - Ner to annotate.
 */

/**
 * @typedef AnnotationPayload
 * @property { Array.<Annotation>} annotations.required - List of annotations.
 * @property { Array.<Relation>} entitiesRelations - List of relations.
 */

/**
 * @typedef AnnotationList
 * @property {UserExtern.model} user - Annotator.
 * @property {string} status - Status of the annotation.
 * @property {string} item - _id of the item.
 * @property {string} value - Value of the annotation.
 * @property {Array.<Zone>} zone - Array of zone.
 * @property {object} task - Classification.
 * @property {Ner.model} ner - Ner.
 */

const init = (router: Router) => {
  /**
   * List project items ( min project role : user ) - Check List Model for pagination details.
   * @route GET /projects/:projectId/items
   * @group items
   * @security Bearer
   * @returns { List.model } 200 - Success.
   * @returns {AnnottoError.model} 401 - Not authorized.
   * @returns {AnnottoError.model} 403 - Forbidden.
   * @returns {AnnottoError.model} 500 - Server Error.
   */

  router.get('/projects/:projectId/items', auth.isAuthenticated({ projectRole: 'user' }), items.index)

  /**
   * List items tags ( min project role : user ).
   * @route GET /projects/:projectId/items/tags
   * @group items
   * @security Bearer
   * @returns { Array.<string> } 200 - Success.
   * @returns {AnnottoError.model} 401 - Not authorized.
   * @returns {AnnottoError.model} 403 - Forbidden.
   * @returns {AnnottoError.model} 500 - Server Error.
   */

  router.get('/projects/:projectId/items/tags', auth.isAuthenticated({ projectRole: 'user' }), items.getTags)

  /**
   * Get a project item to annotate ( min project role : user ).
   *
   * FindOne sorted by descending seenAt ( an unseen item will be given first, then the oldest seen one ).
   * @route GET /projects/:projectId/items/next
   * @group items
   * @security Bearer
   * @returns { List.model } 200 - Success.
   * @param {string} filterId Pass a filter's id as query param "filterId", to get an item using the filter.
   * @returns {AnnottoError.model} 401 - Not authorized.
   * @returns {AnnottoError.model} 403 - Forbidden.
   * @returns {AnnottoError.model} 404 Filter not found.
   * @returns {AnnottoError.model} 500 Fail to list items.
   */

  router.get('/projects/:projectId/items/next', auth.isAuthenticated({ projectRole: 'user' }), items.next)

  /**
   * Get a specific item by Uuid ( min project role : user ).
   * @route GET /projects/:projectId/items/by-uuid/:itemUuid
   * @group items
   * @security Bearer
   * @returns { Item.model } 200 - Success.
   * @returns {AnnottoError.model} 401 - Not authorized.
   * @returns {AnnottoError.model} 403 - Forbidden.
   * @returns {AnnottoError.model} 404 Item not found.
   * @returns {AnnottoError.model} 500 Fail to get item.
   */

  router.get(
    '/projects/:projectId/items/by-uuid/:itemUuid',
    auth.isAuthenticated({ projectRole: 'user' }),
    items.getByUuid
  )

  /**
   * Get a specific item ( min project role : user ).
   * @route GET /projects/:projectId/items/:itemId
   * @group items
   * @security Bearer
   * @returns { Item.model } 200 - Success.
   * @returns {AnnottoError.model} 401 - Not authorized.
   * @returns {AnnottoError.model} 403 - Forbidden.
   * @returns {AnnottoError.model} 404 Item not found.
   * @returns {AnnottoError.model} 500 Fail to get item.
   */

  router.get('/projects/:projectId/items/:itemId', auth.isAuthenticated({ projectRole: 'user' }), items.getById)

  /**
   * Update a specific item ( min project role : user ).
   * @route PUT /projects/:projectId/items/:itemId
   * @group items
   * @security Bearer
   * @param {ItemPayload.model} item.body.required
   * @returns { Item.model } 200 - Success.
   * @returns {AnnottoError.model} 401 - Not authorized.
   * @returns {AnnottoError.model} 404 Item not found.
   * @returns {AnnottoError.model} 500 Fail to list items.
   */
  router.put(
    '/projects/:projectId/items/:itemId',
    auth.isAuthenticated({ projectRole: 'user' }),
    schemaValidationMiddleware({ schemaBody: updateItemSchema }),
    items.update
  )

  /**
   * Get list of the annotations of an item ( min project role : user ).
   * @route GET /projects/:projectId/items/:itemId/annotations
   * @group items
   * @security Bearer
   * @returns { AnnotationList.model } 200 - Success.
   * @returns { Error } 401 - Not authorized.
   * @returns { Error } 500 Fail to list items.
   */
  router.get(
    '/projects/:projectId/items/:itemId/annotations',
    auth.isAuthenticated({ projectRole: 'user' }),
    items.annotations
  )

  /**
   * Annotate an item from a project ( min project role : user ).
   *
   * Payload for tasks project :
   * ```{"annotations":[{"value":"myClassif"}]}```.
   *
   * Payload for NER project :
   * ```{"annotations":[{"value":"myClassif", "ner": {"start": 0, "end": 10}}]}```.
   *
   * Payload for Zone project :
   * ```{"annotations":[{"value":"myClassif", "zone": [{x:0.1, y:0.2}, {x:0.2, y:0.3}, {x:0.3, y:0.4}]}]}```.
   *
   * -------------.
   *
   * To add relations between entities use optionnal entitiesRelations :
   * ```{"annotations":[{"value":"myClassif", "ner": {"start": 0, "end": 10}}, {"value":"myClassif2", "ner": {"start": 10, "end": 20}}], "entitiesRelations": [{"src": {"value":"myClassif", "ner": {"start": 0, "end": 10}}, "dest": {"value":"myClassif2", "ner": {"start": 10, "end": 20}}, "value": "myRelation"}]}```.
   *
   * @route POST /projects/:projectId/items/:itemId/annotate
   * @group items
   * @security Bearer
   * @param { AnnotationPayload.model} payload.body.required Annotation.
   * @returns { Item.model } 200 - Success Returns item's annotations.
   * @returns {AnnottoError.model} 401 - Not authorized.
   * @returns {AnnottoError.model} 500 Fail to list items.
   */
  router.post(
    '/projects/:projectId/items/:itemId/annotate',
    auth.isAuthenticated({ projectRole: 'user' }),
    schemaValidationMiddleware({ schemaBody: annotationPayloadSchema }),
    items.annotate
  )

  /**
   * Upload a jsonlines file of project items ( min project role : admin ).
   *
   * Jsonlines Image format :
   * ```{ "uuid": "d7bb0128-c478", "datatype": "image", "data": { "url": "myImageUrl"}} ```.
   *
   * Jsonlines Text format :
   * ```{ "uuid": "fbf945d8-4afa", "datatype": "text", "data" : { "text": "Bonjour, je souhaite ... "} }```.
   *
   * @route POST /projects/:projectId/items/upload
   * @group items
   * @security Bearer
   * @param { file } file.body.required File with items.
   * @returns {AnnottoError.model} 401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model} 500 Fail to list items.
   */

  router.post('/projects/:projectId/items/upload', auth.isAuthenticated({ projectRole: 'admin' }), items.itemsUpload)

  /**
   * Upload a jsonlines file of predictions for items uuid ( min project role : admin ).
   *
   * TEXT prediction format
   * ```{ "uuid": "d7bb0128-c478", "annotations": { "myTextCategory": { "entities": [{"value": "myClassificationValueCode", "text": "lorem impsum" }] }}}```.
   *
   * NER prediction format
   * ```{ "uuid": "d7bb0128-c478", "annotations": { "myNERCategory": { "entities": [{"value": "myClassificationValueCode", "start": 34, "end": 38 }] }}}```.
   *
   * Zone prediction format
   * ```{ "uuid": "d7bb0128-c478", "annotations": { "myZoneCategory": { "entities": [ {"value": "myClassificationValueCode", "coords": [ {"x": 0.1, "y": 0.2}, {"x": 0.2, "y": 0.3}, {"x": 0.4, "y": 0.5} ]} ]}}}```.
   *
   * Classification prediction format
   * ```{ "uuid": "d7bb0128-c478", "annotations": { "myClassificationCategory": { "labels": [ {"value":"myClassificationValueCode" , "proba": 0.99 } ] }}}```.
   *
   * Zone AND Classification format
   * ```{ "uuid": "526ccc0c-ebab", "annotations": { "myClassificationCategory": { "labels": [ {"value": "myClassificationValueCode", "proba": 0.99 } ] }}, "zone": { "myZoneCategory": { "entities": [ {"value": "bbox_name", "coords": [{"x": 0.1, "y": 0.2}, {"x": 0.3, "y": 0.4}, {"x": 0.5, "y": 0.6}] } ]}}}```.
   *
   * NER AND Classification format
   * ```{ "uuid": "76030ac8-6d2d", "annotations": { "myClassificationCategory": { "labels":[{"value": "myClassificationValueCode", "proba": 0.99}] }}, "ner": { "myNERCategory": { "entities": [{"value": "name", "start": 23, "end": 33 }] }}}```.
   *
   * @route POST /projects/:projectId/items/predictionsUpload
   * @group items
   * @security Bearer
   * @param { file } file.body.required File with predictions.
   * @returns {AnnottoError.model} 401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model} 500 Fail to list items.
   */
  router.post(
    '/projects/:projectId/items/predictionsUpload',
    auth.isAuthenticated({ projectRole: 'admin' }),
    items.predictionUpload
  )
}

export default init
