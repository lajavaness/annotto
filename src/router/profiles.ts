import { Router } from 'express'
import auth from '../auth/service'
import profiles from '../middleware/profiles'
import { schemaValidationMiddleware } from '../middleware/schema-validation-middleware'
import { rolesSchema } from './validation/profile'

/**
 * @typedef ProfilesList
 * @property {number} count - Number of profiles.
 * @property {number} index - Index of the page.
 * @property {number} pageCount - Page quantity.
 * @property {number} total - Total number of profiles.
 * @property {Array.<Profile>} data - List of profiles.
 */

/**
 * @typedef Profile
 * @property {string} user.required - Profile user-api ID.
 * @property {boolean} role.required - Profile role.
 * @property {string} email.required - Profile email.
 */

const init = (router: Router) => {
  /**
   * List Profiles
   * See confluence Doc for more details : https://lajavaness.atlassian.net/wiki/spaces/AN/pages/2086371385/Permissions.
   * @route GET /profiles
   * @group profiles
   * @security Bearer
   * @returns {ProfilesList.model} 200 - Success.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 Fail to list profiles.
   */
  router.get('/profiles', auth.isAuthenticated({ realmRole: 'admin' }), profiles.index)

  /**
   * Update Profile role.
   * @route PUT /profiles/:id
   * @group profiles
   * @security Bearer
   * @param {string} role.query.required Query param "role" to assign to user ( "user / dataScientist / admin" ).
   * @returns {ProfilesList.model} 200 - Success.
   * @returns {AnnottoError.model}  401 - Not authorized.
   * @returns {AnnottoError.model}  403 - Forbidden.
   * @returns {AnnottoError.model}  500 Fail to list profiles.
   */
  router.put(
    '/profiles/:profileId',
    auth.isAuthenticated({ realmRole: 'admin' }),
    schemaValidationMiddleware({ schemaQuery: rolesSchema }),
    profiles.update
  )
}

export default init
