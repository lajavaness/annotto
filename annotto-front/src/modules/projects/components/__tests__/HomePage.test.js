import { LocationProvider } from '@reach/router'
import { Provider } from 'react-redux'
import { ThemeProvider } from 'styled-components'
import { createStore } from 'redux-dynamic-modules'
import { render } from '@testing-library/react'
import React from 'react'

import theme from '__theme__'

import HomePage from 'modules/projects/components/HomePage'

const initialState = {}
const getStore = (state) => createStore({ initialState: state })
const getInstance = (state = initialState) => (
	<Provider store={getStore(state)}>
		<ThemeProvider theme={theme}>
			<LocationProvider>
				<HomePage />
			</LocationProvider>
		</ThemeProvider>
	</Provider>
)

describe('HomePage', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})
})
