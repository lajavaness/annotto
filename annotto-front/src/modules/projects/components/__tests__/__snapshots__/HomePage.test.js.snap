// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-cvJIAo kRaRtU"
    data-testid="__loader__"
  >
    <div
      class="ant-spin ant-spin-lg ant-spin-spinning"
    >
      <span
        class="ant-spin-dot ant-spin-dot-spin"
      >
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
        <i
          class="ant-spin-dot-item"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
